drivers/input/keyboard/keyboard_gpio.c

```diff
@@ -0,0 +1,421 @@
+
+		gpio_add_callback(gpio->port, &driver->gpio_data);
+	}
+
+	/* Maybe key pressed before driver setup. */
+	k_work_schedule(&data->delayed_work, K_NO_WAIT);
+
+	return 0;
+}
+
+static int kbd_gpio_release(const struct device *dev)
+{
+	const struct kbd_gpio_config *config = dev->config;
+	struct kbd_gpio_data *data = dev->data;
+	struct input_dev *input = data->input;
+
+	for (int i = 0; i < config->num_keys; i++) {
+		const struct gpio_dt_spec *gpio = &config->gpio[i];
+		struct kbd_gpio_driver *driver = &data->driver[i];
+
+		gpio_remove_callback(gpio->port, &driver->gpio_data);
+	}
+
+	k_work_cancel_delayable(&data->delayed_work);
+
+	input_internal_release(input);
+
+	return 0;
+}
+
+static int kbd_gpio_attr_get(const struct device *dev,
+			enum input_attr_type type, union input_attr_data *attr)
+{
+	struct kbd_gpio_data *data = dev->data;
+	struct input_dev *input = data->input;
+	int retval = 0;
+
+	retval = input_internal_attr_get(input, type, attr);
+
+	return retval;
+}
+
+static int kbd_gpio_attr_set(const struct device *dev,
+			enum input_attr_type type, union input_attr_data *attr)
+{
+	struct kbd_gpio_data *data = dev->data;
+	struct input_dev *input = data->input;
+	int retval = 0;
+
+	retval = input_internal_attr_set(input, type, attr);
+
+	return retval;
+}
+
+static int kbd_gpio_event_read(const struct device *dev, struct input_event *event)
+{
+	struct kbd_gpio_data *data = dev->data;
+	struct input_dev *input = data->input;
+	int retval = 0;
+
+	retval = input_internal_event_read(input, event);
+
+	return retval;
+}
+
+static int kbd_gpio_event_write(const struct device *dev, struct input_event *event)
+{
+	struct kbd_gpio_data *data = dev->data;
+	struct input_dev *input = data->input;
+	int retval = 0;
+
+	retval = input_internal_event_write(input, event);
+
+	return retval;
+}
+
+static int kbd_gpio_init(const struct device *dev)
+{
+	const struct kbd_gpio_config *config = dev->config;
+	struct kbd_gpio_data *data = dev->data;
+	int err = 0;
+
+	if (config->num_keys <= 0) {
+		LOG_ERR("%s: no KEYs found (DT child nodes missing)", dev->name);
+		return -ENODEV;
+	}
+
+	data->dev = dev;
+
+	k_work_init_delayable(&data->delayed_work, delayed_work_handler);
+
+	memset(data->driver, 0x00, config->num_keys * sizeof(struct kbd_gpio_driver));
+
+	LOG_DBG("gpio key map:");
+
+	for (int i = 0; i < config->num_keys; i++) {
+		const struct gpio_dt_spec *gpio = &config->gpio[i];
+		const struct key_info_dt_spec *info = &config->info[i];
+		struct kbd_gpio_driver *driver = &data->driver[i];
+
+		driver->pdata = data;
+
+		if (!device_is_ready(gpio->port)) {
+			LOG_WRN("gpio port[%s] is not ready", gpio->port->name);
+			continue;
+		}
+
+		err = gpio_pin_configure_dt(gpio, GPIO_INPUT);
+		if (err != 0) {
+			LOG_WRN("configure extra_flags on gpio[%s %d] fail[%d]",
+					gpio->port->name, gpio->pin, err);
+			continue;
+		}
+
+		err = gpio_pin_interrupt_configure_dt(gpio, GPIO_INT_EDGE_TO_ACTIVE);
```

Comment:
Not all gpio pins support interrupts. This will have limitations and only work on certain hardware configurations.

Also, using edge to active interrupt won't allow deep sleep/wake from deep sleep on nRF52.